src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs

```diff
@@ -0,0 +1,337 @@
+        {
+            var runningPipelines =
+                        await StateManager.GetOrAddAsync<IReliableDictionary<int, IList<ReleasePipelineStatusItem>>>(RunningPipelineDictionaryName);
+            try
+            {
+                HashSet<BuildChannel> buildChannelsToAdd = new HashSet<BuildChannel>(new BuildChannelComparer());
+                using (ITransaction tx = StateManager.CreateTransaction())
+                {
+                    var runningPipelinesEnumerable = await runningPipelines.CreateEnumerableAsync(tx, EnumerationMode.Unordered);
+                    var asyncEnumerator = runningPipelinesEnumerable.GetAsyncEnumerator();
+                    while (await asyncEnumerator.MoveNextAsync(cancellationToken))
+                    {
+                        int buildId = asyncEnumerator.Current.Key;
+                        ConditionalValue<IList<ReleasePipelineStatusItem>> maybeReleaseStatuses = await runningPipelines.TryGetValueAsync(tx, buildId);
+                        if (maybeReleaseStatuses.HasValue)
+                        {
+                            bool allFinished = true;
+                            IList<ReleasePipelineStatusItem> releaseStatuses = maybeReleaseStatuses.Value;
+
+                            foreach (ReleasePipelineStatusItem releaseStatus in releaseStatuses)
+                            {
+                                int releaseId = releaseStatus.ReleaseId;
+                                AzureDevOpsClient azdoClient = await GetAzureDevOpsClientForAccount(releaseStatus.PipelineOrganization);
+                                AzureDevOpsRelease release = await azdoClient.GetReleaseAsync(releaseStatus.PipelineOrganization, releaseStatus.PipelineProject, releaseStatus.ReleaseId);
+                                string currentStatus = release.Environments[0].Status;
+                                if (InProgressStatuses.Contains(currentStatus.ToLowerInvariant()))
+                                {
+                                    allFinished = false;
+                                }
+                                else
+                                {
+                                    Logger.LogInformation($"Release {releaseId} finished executing with status: {currentStatus}");
+                                    buildChannelsToAdd.Add(new BuildChannel
+                                    {
+                                        BuildId = buildId,
+                                        ChannelId = releaseStatus.ChannelId
+                                    });
+                                }
+                            }
+
+                            // Stop tracking releases for builds where every release finished.
+                            if (allFinished)
+                            {
+                                await runningPipelines.TryRemoveAsync(tx, asyncEnumerator.Current.Key);
```

Comment:
Changing the implementation of this due to a possible case where this list becomes infinitely long that @alexperovich pointed out.